index.bs

```diff
@@ -292,15 +306,134 @@ the PrivacyCG.
+
+The user agent holds a <dfn>user activation map</dfn> which is a [=map=] of
+[=site=] [=hosts=] to [=moments=].  The [=moments=] represent the most recent
+[=wall clock=] time at which the user activated a top-level document on the
+associated [=host=].
+
+The user agent holds a <dfn>candidate bounce tracking map</dfn> which a [=map=]
+of [=site=] [=hosts=] to [=moments=].  The [=moments=] represent the most recent
+[=wall clock=] time at which a page on the give [=host=] performed an action
+that could indicate bounce tracking took place.
+
+The user agent defines a <dfn>bounce tracking grace period</dfn> constant which
+is a [=duration=].  This represents the length of time after a possible bounce
+tracking event during which the user agent will wait for an interaction before
+deleting a [=host=]'s storage.
+
+The user agent defines a <dfn>bounce tracking activation lifetime</dfn>
+constant which is a [=duration=] representing how long user activations will
+protect a [=host=] from storage deletion.
+
+The user agent defines a <dfn>bounce tracking timer period</dfn> constant which
+is a [=duration=] representing how often to run its timer algorithm.
+
+<p class=note>
+Schemeless site is used as the data structure key because by default cookies
+are sent to both `http://` and `https://` pages on the same domain.
+</p>
 
 <h3 id="bounce-tracking-mitigations-algorithms">Algorithms</h3>
 
 * TODO: Define the steps necessary to detect and store a "bounce".
-* TODO: Define the steps necessary to store user interactions.
-* TODO: Define the steps to analyze information in the data model and delete
-        appropriate sites.
+
+<h4 id="bounce-tracking-mitigations-activation-monkey-patch">User Activation Monkey Patch</h4>
+
+<div algorithm>
+
+To <dfn>record a user activation</dfn> given a [=Document=] |document|, perform
+the following steps:
+
+1. If |document|'s [=node navigable=] is null, then abort these steps.
+1. Let |settings| be |document|'s [=relevant settings object=].
+1. Let |origin| be |settings|' [=environment/top-level origin=].
+1. If |origin| is an [=opaque origin=] then abort these steps.
+1. Let |site| be the result of running [=obtain a site=] given |origin|.
+1. Set [=user activation map=][|site|'s [=host=]] to |settings|'s
+    [=environment settings object/current wall time=].
+
+</div>
+
+Append the following steps to the [=activation notification=] steps in the
+[[HTML#user-activation-processing-model|user activation processing model]]:
+
+1. Run [=record a user activation=] given <var ignore>document</var>.
+
+<h4 id="bounce-tracking-mitigations-timer">Timer</h4>
+
+<div algorithm>
+
+To run the <dfn>bounce tracking timer</dfn> algorithm given a [=moment=] |now|,
+perform the following steps:
+
+1. [=map/For each=] |host| -> |bounceTime| of [=candidate bounce tracking map=]:
+    1. If |bounceTime| + [=bounce tracking grace period=] is less than or equal to
+        |now|, then [=iteration/continue=].
```

Comment:
While a single bounce may be within a single browsing session, the interactions may come in different browsing sessions.  For that reason we can't really use monotonic time.

As a consequence if the user change's their clock significantly it could trigger deletion.  The lifetime of user activation protection is expected to be much longer than the grace period or timer period, though.  So only large changes are likely to cause this to happen.